lib/libcgroup-3.1.0/doc/man/cgcreate.1
.\" Written by Ivana Hutarova Varekova <varekova@redhat.com>

.TH CGCREATE  1 2009-03-15 "Linux" "libcgroup Manual"
.SH NAME
cgcreate \- create new cgroup(s)

.SH SYNOPSIS
\fBcgcreate\fR [\fB-h\fR] [\fB-b\fR] [\fB-c\fR]
[\fB-t\fR <\fItuid>:<tgid\fR>] [\fB-a\fR <\fIagid>:<auid\fR>]
[\fB-f\fR mode] [\fB-d\fR mode]
[\fB-s\fR mode] \fB-g\fR <\fIcontrollers>:<path\fR> [\fB-g\fR ...]

.SH DESCRIPTION
The command creates new cgroup(s) defined by the options
\fB-g\fR.

.TP
.B -a <agid>:<auid>
defines the name of the user and the group which own the
rest of the defined control group’s files. These users are
allowed to set subsystem parameters and create subgroups.
The default value is the same as has the parent cgroup.

.TP
.B -b
ignores the default systemd delegated hierarchy path and
constructs the path of the control groups relative to the
cgroup root hierarchy.

.TP
.B -c, --scope
creates a new systemd scope. The cgroup name provided after the
\fB-g\fR flag must be of the form
\fB<slice-name>.slice/<scope-name>.scope\fR. If the slice
does not exist, systemd will create it. Libcgroup will place an
idle process in the scope's cgroup.procs file unless the \fB-p\fR
flag is provided.

.TP
.B -d, --dperm=mode
sets the permissions of a control groups directory.
The permissions needs to be specified as octal numbers e.g.
\fB-d 775\fR.

.TP
.B -f, --fperm=mode
sets the permissions of the control groups parameters.
The permissions needs to be specified as octal numbers e.g.
\fB-f 775\fR.
The value is not used as given because the current owner's
permissions are used as an umask (so 777 will set group and
others permissions to the owners permissions).

.TP
.B -g <controllers>:<path>
defines control groups to be added.
\fBcontrollers\fR is a list of controllers. Character "*" can be used
as a shortcut for "all mounted controllers".
\fBpath\fR is the relative path to control groups
in the given controllers list. This option can be specified
multiple times.

.TP
.B -h, --help
display this help and exit

.TP
.B -p, --pid=pid
moves the provided \fBpid\fR into the
\fB<slice-name>.slice/<scope-name>.scope\fR.  Must be used in
conjunction with \fB-c\fR.

.TP
.B -s, --tperm=mode
sets the permissions of the control group tasks file.
The permissions needs to be specified as octal numbers e.g.
\fB-s 775\fR.
The value is not used as given because the current owner's
permissions are used as an umask (so 777 will set group and
others permissions to the owners permissions).

.TP
.B -S, --setdefault
sets the scope specified via the \fB-c\fR option as the default
libcgroup scope.  This default scope is used by libcgroup tools
to build up the cgroup path.  See the \fB-b\fR option.

.TP
.B -t <tuid>:<tgid>
defines the name of the user and the group, which owns tasks
file of the defined control group. I.e. this user and members
of this group have write access to the file.
The default value is the same as has the parent cgroup.

.SH ENVIRONMENT VARIABLES
.TP
.B CGROUP_LOGLEVEL
controls verbosity of the tool. Allowed values are \fBDEBUG\fR,
\fBINFO\fR, \fBWARNING\fR or \fBERROR\fR.

.SH EXAMPLES
.TP
.B cgcreate -g *:student devices:teacher
create control group student in all mounted hierarchies and create
control group teacher in hierarchy containing controller devices.




.SH SEE ALSO
cgrules.conf (5)
cgexec (1)
cgclassify (1)


ChangeLog
DOMjudge Programming Contest Judging System

Version 8.4.0DEV
---------------------------

Version 8.3.0 - 31 May 2024
---------------------------
 - [security] Close metadata file descriptor for the child in runguard.
 - Document that minimum PHP version is now 8.1.0.
 - Upgrade Symfony to 6.2 and upgrade other library dependencies.
 - Show link to problem text on jury problem overview and contest detail pages.
 - Add label to teams to be in sync with ICPC CCS spec.
 - Add feedback/teammessage.txt for automated teamfeedback, see: #2016.
 - Clarifications can now use markdown syntax.
 - Upgrade Bootstrap to 5.2 and dependent components.
 - Bug fixed to correctly select current user during configuration of installation.
 - Properly fix removal of problem/submission data.
 - Filter away balloons of jury members in balloon runner interface.
 - Fix upgrade path for executables, see: #1826.
 - Collect compiler and runner versions before compilation, useful to compare
   judgehosts.
 - Various improvements to shadow config used at ICPC World Finals.
 - Show the historic timelimit which was used when judging the submission.
 - Fixes to the documentation and import of problems with @EXPECTED_RESULTS@
   tags.
 - Allow filtering on affilition in submission page for jury for multisite
   contests.
 - Add preview of the contest problemset to the jury page before contest
   start.
 - Add optional Adminer for easy database access.
 - Fix medal awards when skipping categories.
 - Add direct button to public page for registering when enabled.
 - Scale batch size of judge tasks dynamically.
 - Rename room to location for teams.
 - Add option to upload contest problemset document.
 - Improve on the advice for rejuding on config changes, only when relevant
   advice to start a rejudging, similar for updating the scoreboard cache.
 - For interactive problems: add explicit statement to when one of sides of
   the interactions ran closed their output.
 - Allow importing a (generic) contest warning which is displayed on the scoreboard.
 - Added option in the dj_setup_database to make dumps, useful before major
   upgrades and during the contest before problem imports.
 - Fixed some bugs to become OpenAPI spec compliant again.
 - Prevent a race condition leading to two submissions receiving the same
   colour balloon.
 - Compile executables with `-std=gnu++20` now.
 - Record the testcase directory for judging runs, this makes investigating
   easier.
 - Set default C/C++ standards in configure script, we intend to ship all
   scripts free of warnings/errors.
 - Respect basic authentication in the URL for the import-contest when .netrc
   exists.
 - Fix CORS header for API to make extending with other tools easier.
 - Fix eventfeed when SQL commits happen out of order to send all events.
 - Lazy load the team pictures for the scoreboard, to prevent spamming access
   log with HTTP404s
 - Make TLE detection more robust for interactive problems.
 - Added option for ICPC online scoreboard with no rank displays.

Version 8.2.0 - 6 March 2023
----------------------------
 - Various bugfixes & UI improvements.
 - Add support for Rust as submission language.
 - Add teamphotos to scoreboard zip.
 - Add jury option to to download ZIP with all problem samples, attachments
   and problem statement.
 - Extend team import with optional room and other API extensions, see the API
   doc for further details.
 - Drop support for shadowing a single contest from multiple sources.
 - Log account logins via API separately.
 - Add support for PHP v8.2.
 - Upgraded library dependencies.
 - Removed output of detected entry point to teams.
 - Added option to add custom apt repositories to judge chroot.
 - Added CLI script to configure the domserver with a yaml file.
 - Changed activation behaviour of contests imported via UI to be directly
   available.
 - New option added to reset login status of teams between contests.
 - Highlight submissions which were judged with compiler flags which since
   have changed. Useful when you find wrong flags during the contest.
 - Handle network errors for judgedaemon when domserver is unavailable.
 - Added super lazy mode useful for ICPC analyst instances. Here nothing is
   judged unless manually requested.
 - Split problem import messages in info/warning/danger, to make sure errors
   are not missed when importing.
 - Store user login sessions in database.
 - Contest import CLI script is now rewritten in python.
 - The example test problems are now imported via the contest package format.

Version 8.1.0 - 1 June 2022
---------------------------
 - Document that minimum PHP version is now 7.4.0.
 - Add support for PHP 8.1.
 - Upgrade Symfony to 5.4 LTS and upgrade other library dependencies.
 - Replace calls to deprecated `strftime` with `date`. Note: if you have
   configured a custom `time_format`, you need to update it to the format
   supported by `date`.
 - Dropped the presentation-error verdict, databases from before DOMjudge 5.0
   will be altered to replace the verdict with wrong-answer.
 - Changed API import to give admin users a team.
 - Balloon users can now be created via the API.
 - Shadow mode will now fix missing data by reading from the primary CCS.
 - Fixed cgroup naming for docker to prevent sharing the same container.
 - Removed python2 in the chroot, remove python3 in favor of pypy3 in chroot.
 - Added support for sentry for logging.
 - Added default external IDs on all places.
 - Fixed medals on scoreboard for filtered scoreboards.
 - Dropped old balloons CLI.
 - Reimplementation of runpipe for interactive problems.
 - Individual files can now be removed from executables.
 - Added mysql socket option, for when port/IP is not used.
 - Added support for accounts.{yaml,json} for import script.
 - Added support for judgeremaining for problems.
 - Add option to delete existing testcases/attachments/other data when
   importing a problem.

lib/libcgroup-3.1.0/src/tools/cgcreate.c
// SPDX-License-Identifier: LGPL-2.1-only
/**
 * Copyright Red Hat, Inc. 2009
 *
 * Authors:	Ivana Hutarova Varekova <varekova@redhat.com>
 */

#include "tools-common.h"

#include <libcgroup.h>
#include <libcgroup-internal.h>

#include <getopt.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <errno.h>
#include <pwd.h>
#include <grp.h>

#include <sys/types.h>

/*
 * Display the usage
 */
static void usage(int status, const char *program_name)
{
	if (status != 0) {
		err("Wrong input parameters, try %s -h for more information.\n", program_name);
		return;
	}

	info("Usage: %s [-h] [-f mode] [-d mode] [-s mode] ", program_name);
	info("[-t <tuid>:<tgid>] [-a <agid>:<auid>] -g <controllers>:<path> [-g ...]\n");
	info("Create control group(s)\n");
	info("  -a <tuid>:<tgid>		Owner of the group and all its files\n");
	info("  -d, --dperm=mode		Group directory permissions\n");
	info("  -f, --fperm=mode		Group file permissions\n");
	info("  -g <controllers>:<path>	Control group which should be added\n");
	info("  -h, --help			Display this help\n");
	info("  -s, --tperm=mode		Tasks file permissions\n");
	info("  -t <tuid>:<tgid>		Owner of the tasks file\n");
#ifdef WITH_SYSTEMD
	info("  -b				Ignore default systemd");
	info("delegate hierarchy\n");
	info("  -c, --scope			Create a delegated systemd scope\n");
	info("  -p, --pid=pid			Task pid to use to create systemd ");
	info("scope\n");
	info("  -S, --setdefault		Set this scope as the default scope ");
	info("delegate hierarchy\n");
#endif
}

#ifdef WITH_SYSTEMD
static int create_systemd_scope(struct cgroup * const cg, const char * const prog_name,
				int set_default, pid_t pid)
{
	struct cgroup_systemd_scope_opts opts;
	char slice[FILENAME_MAX];
	char *scope;
	int ret;
	int len;

	ret = cgroup_set_default_scope_opts(&opts);
	if (ret)
		return ret;

	opts.pid = pid;

	ret = cgroup_create_scope2(cg, 0, &opts);
	if (!ret && set_default) {
		scope = strstr(cg->name, "/");
		if (!scope) {
			err("%s: Invalid scope name %s, expected <slice>/<scope>\n",
			    prog_name, cg->name);
			ret = ECGINVAL;
			goto err;
		}
		len = strlen(cg->name) - strlen(scope);
		strncpy(slice, cg->name, FILENAME_MAX - 1);
		slice[len] = '\0';
		scope++;

		ret = cgroup_write_systemd_default_cgroup(slice, scope);
		/*
		 * cgroup_write_systemd_default_cgroup() returns 0 on failure
		 */
		if (ret == 0) {
			err("%s: failed to write default %s/%s to ",
			    prog_name, slice, scope);
			err("/var/run/libcgroup/systemd\n");
			ret = ECGINVAL;
			goto err;
		}

		/*
		 * the default was successfully set.  override the return of "1" back to
		 * the usual "0" on success.
		 */
		ret = 0;
        }

err:
	return ret;
}
#else
static int create_systemd_scope(struct cgroup * const cg, const char * const prog_name,
				int set_default, pid_t pid)
{
	return ECGINVAL;
}
#endif /* WITH_SYSTEMD */

int main(int argc, char *argv[])
{
	static struct option long_opts[] = {
		{"help",	      no_argument, NULL, 'h'},
		{"task",	required_argument, NULL, 't'},
		{"admin",	required_argument, NULL, 'a'},
		{"",		required_argument, NULL, 'g'},
		{"dperm",	required_argument, NULL, 'd'},
		{"fperm",	required_argument, NULL, 'f'},
		{"tperm",	required_argument, NULL, 's'},
#ifdef WITH_SYSTEMD
		{"scope",	      no_argument, NULL, 'c'},
		{"setdefault",	      no_argument, NULL, 'S'},
		{"pid",		required_argument, NULL, 'p'},
#endif /* WITH_SYSTEMD */
		{0, 0, 0, 0},
	};

	uid_t tuid = CGRULE_INVALID, auid = CGRULE_INVALID;
	gid_t tgid = CGRULE_INVALID, agid = CGRULE_INVALID;

	int ignore_default_systemd_delegate_slice = 0;
	int set_default_scope = 0;
	int create_scope = 0;
	pid_t scope_pid = -1;

	struct cgroup_group_spec **cgroup_list;
	struct cgroup_controller *cgc;
	struct cgroup *cgroup;

	/* approximation of max. numbers of groups that will be created */
	int capacity = argc;

	/* permission variables */
	mode_t tasks_mode = NO_PERMS;
	mode_t file_mode = NO_PERMS;
	mode_t dir_mode = NO_PERMS;
	int filem_change = 0;
	int dirm_change = 0;

	int ret = 0;
	int i, j;
	int c;

	/* no parametr on input */
	if (argc < 2) {
		usage(1, argv[0]);
		exit(EXIT_BADARGS);
	}

	cgroup_list = calloc(capacity, sizeof(struct cgroup_group_spec *));
	if (cgroup_list == NULL) {
		err("%s: out of memory\n", argv[0]);
		ret = -1;
		goto err;
	}

#ifdef WITH_SYSTEMD
	/* parse arguments */
	while ((c = getopt_long(argc, argv, "a:t:g:hd:f:s:bcp:S", long_opts, NULL)) > 0) {
		switch (c) {
		case 'b':
			ignore_default_systemd_delegate_slice = 1;
			break;
		case 'c':
			create_scope = 1;
			break;
		case 'p':
			scope_pid = atoi(optarg);
			if (scope_pid <= 1) {
				err("%s: Invalid pid %s\n", argv[0], optarg);
				ret = EXIT_BADARGS;
				goto err;
			}
			break;
		case 'S':
			set_default_scope = 1;
			break;
#else
	while ((c = getopt_long(argc, argv, "a:t:g:hd:f:s:", long_opts, NULL)) > 0) {
		switch (c) {
#endif
		case 'h':
			usage(0, argv[0]);
			ret = 0;
			goto err;
		case 'a':
			/* set admin uid/gid */
			if (parse_uid_gid(optarg, &auid, &agid, argv[0]))
				goto err;
			break;
		case 't':
			/* set task uid/gid */
			if (parse_uid_gid(optarg, &tuid, &tgid, argv[0]))
				goto err;
			break;
		case 'g':
			ret = parse_cgroup_spec(cgroup_list, optarg, capacity);
			if (ret) {
				err("%s: cgroup controller and path parsing failed (%s)\n",
				    argv[0], argv[optind]);
				ret = EXIT_BADARGS;
				goto err;
			}
			break;
		case 'd':
			dirm_change = 1;
			ret = parse_mode(optarg, &dir_mode, argv[0]);
			if (ret)
				goto err;
			break;
		case 'f':
			filem_change = 1;
			ret = parse_mode(optarg, &file_mode, argv[0]);
			if (ret)
				goto err;
			break;
		case 's':
			filem_change = 1;
			ret = parse_mode(optarg, &tasks_mode, argv[0]);
			if (ret)
				goto err;
			break;
		default:
			usage(1, argv[0]);
			ret = EXIT_BADARGS;
			goto err;
		}
	}

#ifdef WITH_SYSTEMD
	if (ignore_default_systemd_delegate_slice && create_scope) {
		err("%s: \"-b\" and \"-c\" are mutually exclusive\n", argv[0]);
		ret = EXIT_BADARGS;
		goto err;
	}

	if (set_default_scope && !create_scope) {
		err("%s: \"-S\" requires \"-c\" to be provided\n", argv[0]);
		ret = EXIT_BADARGS;
		goto err;
	}

	if (!create_scope && scope_pid != -1) {
		err("%s: \"-p\" requires \"-c\" to be provided\n", argv[0]);
		ret = EXIT_BADARGS;
		goto err;
	}
#endif

	/* no cgroup name */
	if (argv[optind]) {
		err("%s: wrong arguments (%s)\n", argv[0], argv[optind]);
		ret = EXIT_BADARGS;
		goto err;
	}

	/* initialize libcg */
	ret = cgroup_init();
	if (ret) {
		err("%s: libcgroup initialization failed: %s\n", argv[0], cgroup_strerror(ret));
		goto err;
	}

	/* this will always be false if WITH_SYSTEMD is not defined */
	if (!create_scope && !ignore_default_systemd_delegate_slice)
		cgroup_set_default_systemd_cgroup();

	/* for each new cgroup */
	for (i = 0; i < capacity; i++) {
		if (!cgroup_list[i])
			break;

		/* create the new cgroup structure */
		cgroup = cgroup_new_cgroup(cgroup_list[i]->path);
		if (!cgroup) {
			ret = ECGFAIL;
			err("%s: can't add new cgroup: %s\n", argv[0], cgroup_strerror(ret));
			goto err;
		}

		/* set uid and gid for the new cgroup based on input options */
		ret = cgroup_set_uid_gid(cgroup, tuid, tgid, auid, agid);
		if (ret)
			goto err;

		/* add controllers to the new cgroup */
		j = 0;
		while (cgroup_list[i]->controllers[j]) {
			if (strcmp(cgroup_list[i]->controllers[j], "*") == 0) {
				/* it is meta character, add all controllers */
				ret = cgroup_add_all_controllers(cgroup);
				if (ret != 0) {
					ret = ECGINVAL;
					err("%s: can't add all controllers\n", argv[0]);
					cgroup_free(&cgroup);
					goto err;
				}
			} else {
				cgc = cgroup_add_controller(cgroup,
					cgroup_list[i]->controllers[j]);
				if (!cgc) {
					ret = ECGINVAL;
					err("%s: controller %s can't be add\n", argv[0],
					    cgroup_list[i]->controllers[j]);
					cgroup_free(&cgroup);
					goto err;
				}
			}
			j++;
		}

		/* all variables set so create cgroup */
		if (dirm_change | filem_change)
			cgroup_set_permissions(cgroup, dir_mode, file_mode, tasks_mode);

		if (create_scope) {
			ret = create_systemd_scope(cgroup, argv[0], set_default_scope, scope_pid);
		} else {
			ret = cgroup_create_cgroup(cgroup, 0);
		}
		if (ret) {
			err("%s: can't create cgroup %s: %s\n", argv[0], cgroup->name,
			    cgroup_strerror(ret));
			cgroup_free(&cgroup);
			goto err;
		}
		cgroup_free(&cgroup);
	}

err:
	if (cgroup_list) {
		for (i = 0; i < capacity; i++) {
			if (cgroup_list[i])
				cgroup_free_group_spec(cgroup_list[i]);
		}
		free(cgroup_list);
	}

	return ret;
}
